src/analysis/retail/druid/balance/modules/spells/CelestialAlignment.tsx

```diff
@@ -0,0 +1,128 @@
+  /** Either CA or Incarnation depending on talent */
+  cdSpell: Spell;
+  /** The duration of current version of spell */
+  cdDuration: number;
+  /** Tracker for each CA cast */
+  caTrackers: CaCast[] = [];
+
+  constructor(options: Options) {
+    super(options);
+    this.active = this.selectedCombatant.hasTalent(TALENTS_DRUID.CELESTIAL_ALIGNMENT_TALENT);
+
+    this.cdSpell = cdSpell(this.selectedCombatant);
+    this.cdDuration = cdDuration(this.selectedCombatant);
+
+    this.addEventListener(Events.cast.by(SELECTED_PLAYER).spell(this.cdSpell), this.onCdUse);
+  }
+
+  onCdUse(event: CastEvent) {
+    this.caTrackers.push({
+      startTimestamp: event.timestamp,
+      endTimestamp: Math.min(event.timestamp + this.cdDuration, this.owner.fight.end_time),
+      activeTimePercentage: undefined,
+    });
+  }
+
+  _fillActiveTimePercentages() {
+    this.caTrackers.forEach((cast) => {
+      if (cast.activeTimePercentage === undefined) {
+        cast.activeTimePercentage = this.deps.alwaysBeCasting.getActiveTimePercentageInWindow(
+          cast.startTimestamp,
+          cast.endTimestamp,
+        );
+      }
+    });
+  }
+
+  get guideCastBreakdown(): JSX.Element {
+    this._fillActiveTimePercentages();
+
+    const explanation = (
+      <p>
+        <strong>
+          <SpellLink spell={this.cdSpell} />
+        </strong>{' '}
+        is our primary damage cooldown. It's best used as soon as it's available, but can be held to
+        ensure you'll have full target uptime during its duration (don't use it when it will be
+        interrupted by a fight mechanic).
+      </p>
+    );
+
+    const data = (
+      <div>
+        <strong>Per-Cast Breakdown</strong>
+        <small> - click to expand</small>
+        {this.caTrackers.map((cast, idx) => {
+          const header = (
+            <>
+              @ {this.owner.formatTimestamp(cast.startTimestamp)} &mdash;{' '}
+              <SpellLink spell={this.cdSpell} />
+            </>
+          );
+
+          const activeTimePercentage = cast.activeTimePercentage ?? 0;
+          let percentActivePerf;
+          if (activeTimePercentage >= PERFECT_INCARN_ACTIVE) {
+            percentActivePerf = QualitativePerformance.Perfect;
+          } else if (activeTimePercentage >= GOOD_INCARN_ACTIVE) {
+            percentActivePerf = QualitativePerformance.Good;
+          } else if (activeTimePercentage >= OK_INCARN_ACTIVE) {
+            percentActivePerf = QualitativePerformance.Ok;
+          } else {
+            percentActivePerf = QualitativePerformance.Fail;
+          }
+
+          const checklistItems: CooldownExpandableItem[] = [];
+          checklistItems.push({
+            label: <>Stay Active!</>,
+            result: <PerformanceMark perf={percentActivePerf} />,
+            details: <>({formatPercentage(cast.activeTimePercentage ?? 0, 0)}% active time)</>,
+          });
+          // TODO more checks?
```

Comment:
i do like this check! especially for specs with really important CD windows, active time within the CD window seems very useful

i know a lot of the rules from SL incarn were venthyr-specific, but might be something about resources spent?

Reply:
I don't think raw spent AsP is a helpful metric, but I could also do a specific "don't overcap" rule within Incarn. Gonna check more with the Balance TC community about other things that could be added here